### PR TITLE
Add BaseCallable trait type

### LINKED ISSUE #774: Add BaseCallable
Now that `Callable` has a fast validator, it's hard to subclass. Following the usual pattern, we should refactor to introduce a `BaseCallable` trait type that can be subclassed.

(There may be scope for re-evaluating that "usual pattern", but that shouldn't happen before the 6.0 release, and would touch a lot more than just "Callable".)

### PR DESCRIPTION
**Pull request by shoeb-github**
closes #774.

### CODE REVIEW

**Comment by shoeb-github** [comment]
I added the suggested test case.

**Comment by mdickinson** [comment]
Thank you! Please could we also have a test that covers basic use of `BaseCallable`: right now, the new code isn't fully covered by existing tests. (In general, when adding new functionality, I recommend running the test suite under the `coverage` tool: that way, you can get a sense of which parts of the new code are covered by the test suite. It's not a requirement to have 100% coverage, but common code paths should be tested.)

File: traits/tests/test_callable.py [hunk 1]
```diff
@@ -57,3 +58,29 @@ def test_rejects_int(self):
...
         self.assertIn(
             "must be a callable value", str(exception_context.exception))
+
+
+class TestBaseCallable(unittest.TestCase):
+
+    def test_override_validate(self):
+
+        class ZeroArgsCallable(BaseCallable):
+
```
Comment (mdickinson) [inline traits/tests/test_callable.py]: It would be good to have an explanatory docstring on this class.

File: traits/tests/test_callable.py [hunk 2]
```diff
@@ -57,3 +58,29 @@ def test_rejects_int(self):
...
+            def validate(self, object, name, value):
+                if callable(value):
+                    sig = Signature.from_callable(value)
```
Comment (mdickinson) [inline traits/tests/test_callable.py]: Why not just `inspect.signature(value)`?

**Comment by shoeb-github** [comment]
Thanks for the suggestion to run tests under `coverage`. I will add more tests and explanatory docstring to `BaseCallable`.

> Why not just inspect.signature(value)?
No particular reason. I will change this.

**Pull request merged**